File: 2023/03/03.py
from sys import stdin
from collections import defaultdict

lines = stdin.read().strip().split('\n')
gear_to_num = defaultdict(list)
s1 = 0

def is_valid(y, x):
    return 0 <= y < len(lines) and 0 <= x < len(lines[0])

for y in range(len(lines)):
    for x_start in range(len(lines[y])):
        x_end = x_start
        if lines[y][x_start].isdigit() and (not is_valid(y, x_start-1) or not lines[y][x_start-1].isdigit()):
            while is_valid(y, x_end+1) and lines[y][x_end+1].isdigit():
                x_end += 1
            good = False
            for yd in range(y-1, y+2):
                for xd in range(x_start-1, x_end+2):
                    if is_valid(yd, xd) and lines[yd][xd] not in '.0123456789':
                        good = True
                        if lines[yd][xd] in '*':
                            gear_to_num[(yd, xd)].append(int(lines[y][x_start:x_end+1]))

            s1 += good * int(lines[y][x_start:x_end+1])

s2 = sum(n[0] * n[1] for n in gear_to_num.values() if len(n) == 2)
print(s1, s2, sep="\n")

File: 2023/03/03.rs
use std::{io::{stdin, Read}, collections::HashMap};
use regex::Regex;

fn is_valid(y: i32, x: i32, lines: &Vec<&str>) -> bool {
    (0..lines.len() as i32).contains(&y) && (0..lines[y as usize].len() as i32).contains(&x)
}

fn main() {
    let mut input = String::new();
    stdin().read_to_string(&mut input).unwrap();
    let lines = input.trim().split("\n").collect::<Vec<_>>();

    let mut gear_to_nums = HashMap::new();

    let mut s1 = 0;
    let mut num;
    for (y, line) in lines.iter().enumerate() {
        let re = Regex::new(r"\d+").unwrap();
        for cap in re.captures_iter(line).map(|g| g.get(0).unwrap()) {
            num = line[cap.start()..cap.end()].parse::<i32>().unwrap();
            let mut good = false;
            for ya in y as i32 - 1..y as i32 + 2 {
                for xa in cap.start() as i32 - 1..cap.end() as i32 + 1 {
                    if is_valid(ya, xa, &lines) {
                        let c = lines[ya as usize].chars().nth(xa as usize).unwrap();
                        good |= !c.is_ascii_digit() && c != '.';
                        if c == '*' {
                            gear_to_nums.entry((ya, xa)).or_insert_with(Vec::new).push(num);
                        }
                    }
                }
            }
            if good {
                s1 += num;
            }
        }
    }
    let s2: i32 = gear_to_nums
        .values()
        .filter(|v| v.len() == 2)
        .map(|v| v[0] * v[1])
        .sum();
    println!("{}\n{}", s1, s2);
}